### PR TITLE
ISPN-13985 Test AsyncCache new API

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-13985
https://issues.redhat.com/browse/ISPN-13994

A follow-up of #10196 adding some tests to the `AsyncCache`, once that one is integrated I'll rebase.
I left out all things that returns a `Flow` and I'll add into a subsequent PR for the `MutinyCache`.
Some small fixes were necessary, and I need to think on a better test abstraction to ease development for the next tests.

Here the diff: https://github.com/wburns/infinispan/compare/ISPN-9893_HR_4.0...jabolina:infinispan:ISPN-13985

### CODE REVIEW

File: client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java
Comment (wburns) [inline client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java:280]: So this is a separate bug? We should probably log a separate JIRA for this.

**Comment by wburns** [comment]
Changes LGTM so far. Put on hold, until the other PR is integrated.

**Comment by wburns** [comment]
Looks like there is a build failure on CI. The SecureCacheImpl class needs to be updated to rename the methods.

File: client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java
Comment (jabolina) [reply client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java:280]: Yes. There are a few other small things I've changed too. I can open a JIRA for them.

**Comment by jabolina** [comment]
Just rebased. Prefer to create a separate JIRA for the bugs and leave this one only with the testing stuff?

**Comment by wburns** [comment]
> Just rebased. If you prefer, I can create a separate JIRA for the bugs and leave this one only with the testing stuff?

Just create a separate JIRA with a separate commit in this PR is fine.

File: client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java
Comment (jabolina) [reply client/hotrod/src/main/java/org/infinispan/hotrod/configuration/HotRodConfigurationBuilder.java:280]: Created and added a commit.

File: client/hotrod/src/test/java/org/infinispan/hotrod/test/AbstractSingleHotRodServerTest.java
Comment (wburns) [inline client/hotrod/src/test/java/org/infinispan/hotrod/test/AbstractSingleHotRodServerTest.java:20]: We should probably rename this to Async somehow as it contains the AsyncCache.

File: core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java
Comment (wburns) [inline core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java:41]: Looks like this needs another JIRA as well?

Comment (jabolina) [reply core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java:41]: Ooops, yeah. I'll omit this one for now since I'll probably come back to the test suite later,

File: client/hotrod/src/test/java/org/infinispan/hotrod/test/AbstractSingleHotRodServerTest.java
Comment (jabolina) [reply client/hotrod/src/test/java/org/infinispan/hotrod/test/AbstractSingleHotRodServerTest.java]: Renamed to a mouthful of a name.

File: core/src/main/java/org/infinispan/AdvancedCache.java
Comment (wburns) [inline core/src/main/java/org/infinispan/AdvancedCache.java:349]: We should remove the default method now.

File: core/src/main/java/org/infinispan/AdvancedCache.java
Comment (wburns) [inline core/src/main/java/org/infinispan/AdvancedCache.java:414]: Same, remove default method.

File: core/src/main/java/org/infinispan/AdvancedCache.java
Comment (wburns) [inline core/src/main/java/org/infinispan/AdvancedCache.java:565]: Same, remove default method.

File: core/src/main/java/org/infinispan/AdvancedCache.java
Comment (wburns) [inline core/src/main/java/org/infinispan/AdvancedCache.java:891]: Same, remove default method.

**Review by wburns — COMMENTED** [review]
LGTM, just a little more cleanup.

**Review by wburns — APPROVED** [review]
LGTM, just waiting on CI

**Comment by jabolina** [comment]
Was missing a serializer for the `ReadCommittedEntry`, just added it now.

**Comment by jabolina** [comment]
I'll put it on hold just for a moment. I'll update the tests to use the metadata and add some missing tests.

File: client/hotrod/src/test/java/org/infinispan/hotrod/util/FlowUtils.java
Comment (wburns) [inline client/hotrod/src/test/java/org/infinispan/hotrod/util/FlowUtils.java:32]:
```suggestion
   public static <R, A> R blockingCollect(Flow.Publisher<T> publisher, Collector<? super T, A, R> collector) {
```
This way the caller can use any type easier. Also if you want you could add the same method for list  that just passes Collectors.toList() to this one.

File: core/src/main/java/org/infinispan/container/entries/ReadCommittedEntry.java
Comment (wburns) [inline core/src/main/java/org/infinispan/container/entries/ReadCommittedEntry.java:375]: I have to admit I am not a fan of making ReadCommittedEntry itself serializable. Especially as all of the sub classes won't work. But I am not sure what the best thing here is to do other than a "view" object that only has the info we care about (key,value, metadata) and only send that across, which would require more changes.

Comment (wburns) [reply core/src/main/java/org/infinispan/container/entries/ReadCommittedEntry.java:375]: An option would be to convert the CacheEntry to an `InternalCacheEntry` just before serialization as that contains everything we would need and all the impls already support externalization afaik.

**Comment by jabolina** [comment]
CI reported a problem with `EmbeddedHotRodCacheListenerTest#testLoadingAndStoringEventsHotRod2` where it expected a `CacheEntryVisitedEvent` after a `putIfAbsent` operation. Updated so we send the event in case we load the entry, this is supposed to have the same semantics as the previous that did a get + put.

**Comment by tristantarrant** [comment]
There is a conflict

**Comment by jabolina** [comment]
Thanks, @tristantarrant. Updated.

File: client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java
Comment (wburns) [inline client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java:270]: I am missing something, shouldn't the version be 1?

File: core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:336]: Sorry can you just explain what happened here?

Comment (jabolina) [reply core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:336]: Previously we did a get + put for a putIfAbsent command, this combination issues a CacheEntryVisited because of the get. With the changes we added, we stopped issuing the event. I did this so we still issue the event when the entry is visited in the putIfAbsent.

File: client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java
Comment (jabolina) [reply client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java:270]: I was expecting that too. I did not check into it, but line 265, with the first replace that returns false, seems to increase the version. Should we register a JIRA or something?

Comment (wburns) [reply client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java:270]: That seems wrong to increment the version if a conditional operation failed. I would say that is a bug.

File: core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java
Comment (jabolina) [reply core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java]: Reverted this stuff and removed the test check.

**Comment by jabolina** [comment]
There is a related failure in `org.infinispan.server.hotrod.HotRodReplicationTest`.

**Comment by karesti** [comment]
@wburns @jabolina which is the status of this one ?

File: client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java
Comment (jabolina) [reply client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java:270]: After looking into it, the behavior is expected:

https://github.com/infinispan/infinispan/blob/d855ad4a2ed126ccc897989e6492039ad5dc1f54/server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java#L288-L294

Comment (wburns) [reply client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java:270]: Alright, I guess technically it doesn't matter as long as they are unique.

**Review by wburns — APPROVED** [review]
LGTM, just waiting for confirmation before merging.

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**
